common/cluster/dynamic_metadata.go

```diff
@@ -0,0 +1,502 @@
+		initialFailoverVersion = d.failoverVersionIncrement
+	}
+
+	d.clusterLock.RLock()
+	defer d.clusterLock.RUnlock()
+	clusterName, ok := d.versionToClusterName[initialFailoverVersion]
+	if !ok {
+		panic(fmt.Sprintf(
+			"Unknown initial failover version %v with given cluster initial failover version map: %v and failover version increment %v.",
+			initialFailoverVersion,
+			d.clusterInfo,
+			d.failoverVersionIncrement,
+		))
+	}
+	return clusterName
+}
+
+func (d *dynamicMetadataImpl) GetFailoverVersionIncrement() int64 {
+	return d.failoverVersionIncrement
+}
+
+func (d *dynamicMetadataImpl) RegisterMetadataChangeCallback(callbackId string, cb CallbackFn) {
+	d.clusterCallbackLock.Lock()
+	d.clusterChangeCallback[callbackId] = cb
+	d.clusterCallbackLock.Unlock()
+
+	oldEntries := make(map[string]*ClusterInformation)
+	newEntries := make(map[string]*ClusterInformation)
+	d.clusterLock.RLock()
+	for clusterName, clusterInfo := range d.clusterInfo {
+		oldEntries[clusterName] = nil
+		newEntries[clusterName] = &ClusterInformation{
+			Enabled:                clusterInfo.Enabled,
+			InitialFailoverVersion: clusterInfo.InitialFailoverVersion,
+			RPCAddress:             clusterInfo.RPCAddress,
+			version:                clusterInfo.version,
+		}
+	}
+	d.clusterLock.RUnlock()
+	cb(oldEntries, newEntries)
+}
+
+func (d *dynamicMetadataImpl) UnRegisterMetadataChangeCallback(callbackId string) {
+	d.clusterCallbackLock.Lock()
+	delete(d.clusterChangeCallback, callbackId)
+	d.clusterCallbackLock.Unlock()
+}
+
+func (d *dynamicMetadataImpl) refreshLoop(ctx context.Context) error {
+	timer := time.NewTicker(refreshInterval)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-timer.C:
+			for err := d.refreshClusterMetadata(ctx); err != nil; err = d.refreshClusterMetadata(ctx) {
+				select {
+				case <-ctx.Done():
+					return nil
+				default:
+					d.logger.Error("Error refreshing remote cluster metadata", tag.Error(err))
+					select {
+					case <-time.After(refreshFailureInterval):
+					case <-ctx.Done():
+						return nil
+					}
+				}
+			}
+		}
+	}
+}
+
+func (d *dynamicMetadataImpl) refreshClusterMetadata(_ context.Context) error {
+	clusterMetadataMap, err := d.listAllClusterMetadata()
+	if err != nil {
+		return err
+	}
+
+	oldEntries := make(map[string]*ClusterInformation)
+	newEntries := make(map[string]*ClusterInformation)
+
+	clusterInfoMap := d.GetAllClusterInfo()
+UpdateLoop:
+	for clusterName, clusterMetadata := range clusterMetadataMap {
+		clusterInfo, ok := clusterInfoMap[clusterName]
+		if !ok {
+			// handle new cluster registry
+			oldEntries[clusterName] = nil
+			newEntries[clusterName] = &ClusterInformation{
+				Enabled:                clusterMetadata.Enabled,
+				InitialFailoverVersion: clusterMetadata.InitialFailoverVersion,
+				RPCAddress:             clusterMetadata.RPCAddress,
+				version:                clusterMetadata.version,
+			}
+			continue UpdateLoop
```

Comment:
we can remove this line and use else case at line 387

Reply:
Good point. Updated